test/common/buffer/buffer_fuzz_test.cc

```diff
@@ -0,0 +1,221 @@
+    const uint32_t previous_length = target_buffer.length();
+    target_buffer.add(source_buffer);
+    ASSERT_EQ(previous_length,
+              target_buffer.search(source_contents.data(), source_contents.size(), 0));
+    break;
+  }
+  case test::common::buffer::Action::kPrependString: {
+    const uint32_t size = clampSize(action.prepend_string());
+    auto string = std::make_unique<std::string>(size, 'a');
+    ctxt.strings_.emplace_back(std::move(string));
+    target_buffer.prepend(absl::string_view(*ctxt.strings_.back()));
+    ASSERT_EQ(0, target_buffer.search(ctxt.strings_.back()->data(), size, 0));
+    break;
+  }
+  case test::common::buffer::Action::kPrependBuffer: {
+    const uint32_t source_index = action.prepend_buffer() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    const std::string source_contents = source_buffer.toString();
+    target_buffer.prepend(source_buffer);
+    ASSERT_EQ(0, target_buffer.search(source_contents.data(), source_contents.size(), 0));
+    break;
+  }
+  case test::common::buffer::Action::kReserveCommit: {
+    const uint32_t previous_length = target_buffer.length();
+    const uint32_t reserve_length = clampSize(action.reserve_commit().reserve_length());
+    constexpr uint32_t reserve_slices = 16;
+    Buffer::RawSlice slices[reserve_slices];
+    uint32_t allocated_slices = target_buffer.reserve(reserve_length, slices, reserve_slices);
+    uint32_t allocated_length = 0;
+    for (uint32_t i = 0; i < allocated_slices; ++i) {
+      ::memset(slices[i].mem_, 'c', slices[i].len_);
+      allocated_length += slices[i].len_;
+    }
+    const uint32_t target_length =
+        std::min(allocated_length, action.reserve_commit().commit_length());
+    uint32_t shrink_length = allocated_length;
+    while (shrink_length > target_length) {
+      ASSERT(allocated_slices > 0);
+      const uint32_t available = slices[allocated_slices - 1].len_;
+      const uint32_t remainder = shrink_length - target_length;
+      if (available > remainder) {
+        slices[allocated_slices - 1].len_ -= remainder;
+        break;
+      }
+      shrink_length -= available;
+      --allocated_slices;
+    }
+    target_buffer.commit(slices, allocated_slices);
+    ASSERT_EQ(previous_length + target_length, target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kCopyOut: {
+    const uint32_t start = action.copy_out().start() % target_buffer.length();
+    uint8_t copy_buffer[2 * 1024 * 1024];
+    const uint32_t length =
+        std::min(static_cast<uint32_t>(target_buffer.length()),
+                 std::min(action.copy_out().length(), static_cast<uint32_t>(sizeof(copy_buffer))));
+    target_buffer.copyOut(start, length, copy_buffer);
+    const std::string contents = target_buffer.toString();
+    ASSERT_EQ(0, ::memcmp(copy_buffer, contents.data(), length));
```

Comment:
Won't this assert fail if `start` is nonzero? `contents` will contain all the content starting from the very beginning of the buffer, but `copy_buffer` won't contain the first `start` bytes.